instrumentation/opentelemetry-instrumentation-aiokafka/src/opentelemetry/instrumentation/aiokafka/utils.py

```diff
@@ -0,0 +1,247 @@
+
+
+ProduceHookT = Optional[Callable[[Span, Tuple, Dict], Awaitable[None]]]
+ConsumeHookT = Optional[
+    Callable[[Span, ConsumerRecord, Tuple, Dict], Awaitable[None]]
+]
+
+HeadersT = List[Tuple[str, Optional[bytes]]]
+
+
+class AIOKafkaContextGetter(textmap.Getter[HeadersT]):
+    def get(self, carrier: HeadersT, key: str) -> Optional[List[str]]:
+        if carrier is None:
+            return None
+
+        for item_key, value in carrier:
+            if item_key == key:
+                if value is not None:
+                    return [value.decode()]
+        return None
+
+    def keys(self, carrier: HeadersT) -> List[str]:
+        if carrier is None:
+            return []
+        return [key for (key, value) in carrier]
+
+
+class AIOKafkaContextSetter(textmap.Setter[HeadersT]):
+    def set(
+        self, carrier: HeadersT, key: Optional[str], value: Optional[str]
+    ) -> None:
+        if carrier is None or key is None:
+            return
+
+        if value is not None:
+            carrier.append((key, value.encode()))
+        else:
+            carrier.append((key, value))
+
+
+_aiokafka_getter = AIOKafkaContextGetter()
+_aiokafka_setter = AIOKafkaContextSetter()
+
+
+def _enrich_span(
+    span: Span,
+    bootstrap_servers: Union[str, List[str]],
+    topic: str,
+    partition: Optional[int],
+) -> None:
+    if not span.is_recording():
+        return
+
+    span.set_attribute(SpanAttributes.MESSAGING_SYSTEM, "kafka")
+    span.set_attribute(SpanAttributes.MESSAGING_DESTINATION, topic)
+
+    if partition is not None:
+        span.set_attribute(SpanAttributes.MESSAGING_KAFKA_PARTITION, partition)
+
+    span.set_attribute(
+        SpanAttributes.MESSAGING_URL, json.dumps(bootstrap_servers)
+    )
+
+
+def _get_span_name(operation: str, topic: str):
+    return f"{topic} {operation}"
+
+
+def _wrap_send(
+    tracer: Tracer, async_produce_hook: ProduceHookT
+) -> Callable[..., Awaitable[None]]:
+    async def _traced_send(
+        func: Callable[..., Awaitable[None]],
+        instance: aiokafka.AIOKafkaProducer,
+        args: Tuple[Any],
+        kwargs: Dict[str, Any],
+    ) -> None:
+        headers = _extract_send_headers(args, kwargs)
+        if headers is None:
+            headers = []
+            kwargs["headers"] = headers
+
+        topic = _extract_send_topic(args, kwargs)
+        bootstrap_servers = _extract_bootstrap_servers(instance.client)
+        partition = await _extract_send_partition(instance, args, kwargs)
+        span_name = _get_span_name("send", topic)
+        with tracer.start_as_current_span(
+            span_name, kind=trace.SpanKind.PRODUCER
+        ) as span:
+            _enrich_span(span, bootstrap_servers, topic, partition)
+            propagate.inject(
+                headers,
+                context=trace.set_span_in_context(span),
+                setter=_aiokafka_setter,
+            )
+            try:
+                if callable(async_produce_hook):
```

Comment:
Similar to [hhtpx](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2823), we should check if the hooks are indeed async.

Reply:
Done